dm/syncer/safe_mode.go

```diff
@@ -43,35 +56,68 @@ func (s *Syncer) enableSafeModeInitializationPhase(tctx *tcontext.Context) {
 		s.safeMode.Add(tctx, 1) // add 1 but has no corresponding -1, so keeps enabled
 		s.tctx.L().Info("enable safe-mode by config")
 	}
-	if s.checkpoint.SafeModeExitPoint() != nil {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
-		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *s.checkpoint.SafeModeExitPoint()))
+	exitPoint := s.checkpoint.SafeModeExitPoint()
+	failpoint.Inject("SafeModeDurationIsZero", func() {
+		exitPoint = nil
+	})
+	if exitPoint != nil {
+		if binlog.CompareLocation(*exitPoint, s.getInitExecutedLoc(), s.cfg.EnableGTID) > 0 {
+			//nolint:errcheck
+			s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
+			s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *exitPoint))
+		} else {
+			s.tctx.L().Info("disable safe-mode because initExecutedLoc equal safeModeExitPoint")
+		}
 	} else {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
-		go func() {
-			defer func() {
-				err := s.safeMode.Add(tctx, -1)
-				if err != nil {
-					// send error to the fatal chan to interrupt the process
-					s.runFatalChan <- unit.NewProcessError(err)
-				}
-				if !s.safeMode.Enable() {
-					s.tctx.L().Info("disable safe-mode after task initialization finished")
+		initPhaseSeconds := s.cfg.SafeModeDuration
+
+		failpoint.Inject("SafeModeInitPhaseSeconds", func(val failpoint.Value) {
+			seconds, _ := val.(string)
+			initPhaseSeconds = seconds
+			s.tctx.L().Info("set initPhaseSeconds", zap.String("failpoint", "SafeModeInitPhaseSeconds"), zap.String("value", seconds))
+		})
+		var duration time.Duration
+		if initPhaseSeconds == "" {
+			duration = time.Second * time.Duration(2*s.cfg.CheckpointFlushInterval)
+		} else {
+			duration, err = time.ParseDuration(initPhaseSeconds)
+			if err != nil {
+				// send error to the fatal chan to interrupt the process
+				s.runFatalChan <- unit.NewProcessError(err)
+				s.tctx.L().Error("enable safe-mode failed due to duration parse failed", zap.Duration("duration", duration))
```

Comment:
```suggestion
				s.tctx.L().Error("enable safe-mode failed due to duration parse failed", zap.Duration("duration", duration))
```

when duration parsing failed, `duration` is the zero value. You should return `initPhaseSeconds`?